src/library/autodj/trackselector.cpp

```diff
@@ -0,0 +1,69 @@
+#include "trackselector.h"
+#include "library/autodj/autodjprocessor.h"
+
+
+
+TrackSelector::TrackSelector(AutoDJProcessor* processor) :
+   QObject(processor),
+   m_processor(processor) {}
+/*
+TrackPointer TrackSelector::getNextTrackFromQueue() {
+    assert(false);
+}
+*/
+
+/* LinearSelector */
+
+LinearSelector::LinearSelector(AutoDJProcessor* processor): TrackSelector(processor) { }
+
+
+TrackPointer LinearSelector::getNextTrackFromQueue() {
+    PlaylistTableModel* model(m_processor->getTableModel());
+
+    while (true) {
+        TrackPointer nextTrack = model->getTrack(model->index(0, 0));
+
+        if (nextTrack) {
+            if (nextTrack->exists()) {
+                return nextTrack;
+            } else {
+                // Remove missing song from auto DJ playlist.
+                model->removeTrack(model->index(0, 0));
+            }
+        } else {
+            // We're out of tracks. Return the null TrackPointer.
+            return nextTrack;
+        }
+    }
+
+}
+
+
+/* RandomSelector */
+RandomSelector::RandomSelector(AutoDJProcessor* processor): TrackSelector(processor) { }
+
+
+TrackPointer RandomSelector::getNextTrackFromQueue() {
+    PlaylistTableModel* model(m_processor->getTableModel());
+
+    while (true) {
+        int trackCount = model->rowCount();
+        int nextPos = qrand()%trackCount;
+        qDebug() << "RandomSelector: chose track row:" << nextPos;
+        TrackPointer nextTrack = model->getTrack(model->index(nextPos, 0));
+
+        if (nextTrack) {
+            if (nextTrack->exists()) {
+                return nextTrack;
+            } else {
+                // Remove missing song from auto DJ playlist.
+                model->removeTrack(model->index(nextPos, 0));
+            }
+        } else {
+            // We're out of tracks. Return the null TrackPointer.
+            return nextTrack;
```

Comment:
The loop could be written in a more readable way:

```
TrackPointer nextTrack;
do {
...
   if (nextTrack) {
      if (nextTrack->exists()) {
        // Found next track
        break;
      } else {
        // Remove missing track from auto DJ playlist.
        model->removeTrack(model->index(nextPos, 0));
        continue;
     }
   }
} while (nextTrack);
return nextTrack;
```
The _continue_ statement is not necessary, but a has nice symmetry with the corresponding _break_ statement in the other branch.